packages/playwright/src/runner/uiMode.ts

```diff
@@ -142,20 +99,71 @@ class TestServer {
       process.stderr.write = this._originalStderrWrite;
     }
   }
+}
 
-  private async _queueListOrRun(method: string, params: any) {
-    if (method === 'list')
-      await this._listTests();
-    if (method === 'run')
-      await this._runTests(params.testIds, params.projects);
+class TestServerDispatcher implements TestServerInterface {
+  private _config: FullConfigInternal;
+  private _globalWatcher: Watcher;
+  private _testWatcher: Watcher;
+  private _testRun: { run: Promise<FullResult['status']>, stop: ManualPromise<void> } | undefined;
+  readonly transport: Transport;
+  private _queue = Promise.resolve();
+
+  constructor(config: FullConfigInternal) {
+    this._config = config;
+    this.transport = {
+      dispatch: (method, params) => (this as any)[method](params),
+      onclose: () => {},
+    };
+    this._globalWatcher = new Watcher('deep', () => this._dispatchEvent('listChanged', {}));
+    this._testWatcher = new Watcher('flat', events => {
+      const collector = new Set<string>();
+      events.forEach(f => collectAffectedTestFiles(f.file, collector));
+      this._dispatchEvent('testFilesChanged', { testFileNames: [...collector] });
+    });
+  }
+
+  async ping() {}
+
+  async open(params: { location: Location }) {
+    // eslint-disable-next-line no-console
+    open('vscode://file/' + params.location.file + ':' + params.location.column).catch(e => console.error(e));
+  }
+
+  async resizeTerminal(params: { cols: number; rows: number; }) {
+    process.stdout.columns = params.cols;
+    process.stdout.rows = params.rows;
+    process.stderr.columns = params.cols;
+    process.stderr.columns = params.rows;
+  }
+
+  async checkBrowsers(): Promise<{ hasBrowsers: boolean; }> {
+    return { hasBrowsers: hasSomeBrowsers() };
+  }
+
+  async installBrowsers() {
+    await installBrowsers();
   }
 
-  private _dispatchEvent(method: string, params?: any) {
-    this._transport?.sendEvent?.(method, params);
+  async listFiles() {
+    try {
+      const runner = new Runner(this._config);
+      return runner.listTestFiles();
+    } catch (e) {
+      const error: TestError = serializeError(e);
+      error.location = prepareErrorStack(e.stack).location;
+      return { projects: [], error };
+    }
+  }
+
+  async listTests(params: { reporter?: string; fileNames: string[]; }) {
+    this._queue = this._queue.then(() => this._innerListTests(params));
```

Comment:
When `_innerListTests` throws, any subsequent operations in the queue will never run. Do we want this behavior? If not, perhaps introduce a `safeEnqueue` wrapper or catch right here? The same applies to all other `_queue` usage.